src/metabase/query_processor/middleware/limit.clj

```diff
@@ -23,14 +24,30 @@
          (qp.util/query-without-aggregations-or-limits? query))
     (update :query assoc :limit max-rows, ::original-limit original-limit)))
 
+(defn- xlsx-export?
+  [& {info :info}]
+  (let [context (:context info)]
+    (= context :xlsx-download)))
+
 (defn determine-query-max-rows
   "Given a `query`, return the max rows that should be returned. This is either:
   1. the output of [[metabase.legacy-mbql.util/query->max-rows-limit]] when called on the given query
   2. [[metabase.query-processor.interface/absolute-max-results]] (a constant, non-nil backstop value)"
   [query]
   (when-not (disable-max-results? query)
-    (or (mbql.u/query->max-rows-limit query)
-        qp.i/absolute-max-results)))
+    (let [user-limit (public-settings/download-row-limit)
+          legacy-max (mbql.u/query->max-rows-limit query)]
+      (cond
+        ;; if we have a maximum from mbql.u, we use it
+        legacy-max                                     legacy-max
+        ;; if the download-row-limit value is below the absolute max, we can use it in all contexts
+        (and user-limit
+             (< user-limit qp.i/absolute-max-results)) user-limit
+        ;; if the download-row-limit value is above the absolute max, we can use it everywhere except for xlsx downloads
+        (and user-limit
+             (not (xlsx-export? query)))               user-limit
+        ;; otherwise we use the absolute max
+        :else                                          qp.i/absolute-max-results))))
```

Comment:
first of all, there's a kind of limitation/bug lurking here, where it's possible for `query->max-rows-limit` to exceed `qp.i/absolute-max-results` and result in an invalid excel file. This was here before though. But in my opinion it should not be possible to download an invalid excel file. Instead we should either raise an exception with a helpful error message, or silently limit the results to `qp.i/absolute-max-results`. We already silently limit to `qp.i/absolute-max-results` if `query->max-rows-limit` returns nil so I guess it we've previously decided silently limiting results is fine, so I think it makes sense to silently limit in all cases when `xlsx-export?` returns true.

assuming that gets fixed, there's a simpler way to implement this, e.g.
```suggestion
    (cond-> (or (mbql.u/query->max-rows-limit query)
                (public-settings/download-row-limit)
                qp.i/absolute-max-results)
      (xlsx-export? query)
      (min qp.i/absolute-max-results))))
```

Reply:
Your code suggestion makes perfect sense, it's much simpler :) 

I don't totally see what you mean about `query->max-rows-limit` though. 
In the non-xlsx cases, we can now have that number be whatever, even more than the absolute max, snice that's dictated by the xlsx limitation.

The code you suggested already guarantees that for xlsx, we clamp the result to the absolute-max-results, so we can't create an invalid xlsx file right, so I don't think anything else needs fixing?

Reply:
> The code you suggested already guarantees that for xlsx, we clamp the result to the absolute-max-results, so we can't create an invalid xlsx file right

Yes, the code I suggested fixes the issue. Sorry my wording wasn't clear. I should have written 

"assuming we want that fixed, there's a simpler way to implement this, e.g."

instead of 

"assuming that gets fixed, there's a simpler way to implement this, e.g."